File: src/bpro/cli/audit.py
"""bpro audit — Run audit (gap detection, gate judgment)."""

from datetime import datetime, timezone
from pathlib import Path

import click
import yaml

from bpro.core.project import get_bpro_dir, load_config, load_specs, load_matrix
from bpro.utils.display import console, print_success, print_warning, print_error, print_info


@click.command("audit")
@click.option("--quick", is_flag=True, default=True, help="Quick audit (file-based, no SLM)")
@click.option("--gate", is_flag=True, help="Run gate judgment (PASS/CONDITIONAL/FAIL)")
def audit(quick: bool, gate: bool) -> None:
    """Run audit: check REQ coverage, test gaps, code changes."""
    bpro_dir = get_bpro_dir()
    if not bpro_dir:
        print_error("Not a bpro project. Run `bpro init` first.")
        raise SystemExit(1)

    reqs = load_specs(bpro_dir)
    if not reqs:
        print_warning("No requirements to audit. Run `bpro plan import` or `bpro snapshot` first.")
        return

    matrix = load_matrix(bpro_dir)
    entries = matrix.get("entries", {}) if matrix else {}
    root = bpro_dir.parent

    results = {"pass": [], "warn": [], "new": [], "stale": [], "fail": []}

    for req in reqs:
        req_id = req["id"]
        status = req.get("status", "")

        if status == "DEPRECATED":
            continue

        entry = entries.get(req_id, {})
        code_refs = entry.get("code_refs", []) or req.get("code_refs", [])
        test_refs = entry.get("test_refs", [])

        # Check code mapping
        has_code = bool(code_refs)
        code_exists = False
        code_changed = False
        if has_code:
            for ref in code_refs:
                ref_path = root / ref
                if ref_path.exists():
                    code_exists = True

        # Check test mapping
        has_test = bool(test_refs)

        # Categorize
        if not has_code:
            if status in ("CONFIRMED", "DRAFT"):
                results["new"].append((req_id, req.get("title", ""), "not implemented"))
            else:
                results["warn"].append((req_id, req.get("title", ""), "no code mapped"))
        elif has_code and not code_exists:
            results["stale"].append((req_id, req.get("title", ""), "code file missing"))
        elif has_code and not has_test:
            results["warn"].append((req_id, req.get("title", ""), "no tests"))
        else:
            results["pass"].append((req_id, req.get("title", ""), "ok"))

    # Display
    console.print()
    total = sum(len(v) for v in results.values())

    for req_id, title, detail in results["pass"]:
        console.print(f"  [green]PASS[/]  {req_id}: {title}")
    for req_id, title, detail in results["warn"]:
        console.print(f"  [yellow]WARN[/]  {req_id}: {title} — {detail}")
    for req_id, title, detail in results["new"]:
        console.print(f"  [blue]TODO[/]  {req_id}: {title} — {detail}")
    for req_id, title, detail in results["stale"]:
        console.print(f"  [red]STALE[/] {req_id}: {title} — {detail}")

    console.print()
    summary = (
        f"  PASS {len(results['pass'])} | "
        f"WARN {len(results['warn'])} | "
        f"TODO {len(results['new'])} | "
        f"STALE {len(results['stale'])}"
    )
    console.print(summary)

    # Gate judgment
    if gate:
        console.print()
        if results["stale"] or results["fail"]:
            console.print("  [red bold]Gate: FAIL[/] — stale or failing items")
        elif results["warn"] or results["new"]:
            console.print("  [yellow bold]Gate: CONDITIONAL PASS[/] — warnings/todos remain")
        else:
            console.print("  [green bold]Gate: PASS[/]")

    # Save audit result
    audit_result = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "summary": {k: len(v) for k, v in results.items()},
        "details": {
            k: [{"id": r[0], "title": r[1], "detail": r[2]} for r in v]
            for k, v in results.items()
        },
    }
    reports_dir = bpro_dir / "reports"
    reports_dir.mkdir(exist_ok=True)
    ts = datetime.now().strftime("%Y%m%d-%H%M%S")
    with open(reports_dir / f"audit-{ts}.yaml", "w") as f:
        yaml.dump(audit_result, f, default_flow_style=False, allow_unicode=True)

    console.print()
    console.print(f"  [dim]Saved to .bpro/reports/audit-{ts}.yaml[/]")

File: src/commands/audit.ts
/**
 * bpro audit — Run audit (gap detection, gate judgment).
 */

import fs from 'node:fs';
import path from 'node:path';
import { Command } from 'commander';
import chalk from 'chalk';
import {
  requireBproDir,
  loadConfig,
  loadSpecs,
  loadMatrix,
  loadModels,
} from '../core/project.js';
import { getConductorAdapter } from '../models/registry.js';
import { AUDIT_SYSTEM_PROMPT, buildAuditPrompt } from '../prompts/audit.js';
import { appendAgentLog } from '../agents/runner.js';
import { printSuccess, printError, printWarning, printInfo, createSpinner } from '../utils/display.js';
import { saveYaml } from '../utils/yaml.js';
import { emitEvent } from '../notifications/index.js';

interface AuditResult {
  id: string;
  title: string;
  detail: string;
}

export const auditCommand = new Command('audit')
  .description('Run audit: check REQ coverage, test gaps, code changes')
  .option('--quick', 'Quick audit (file-based, no model call)', true)
  .option('--gate', 'Run gate judgment (PASS/CONDITIONAL/FAIL)')
  .action(async (opts: { quick?: boolean; gate?: boolean }) => {
    try {
      const bproDir = requireBproDir();
      const reqs = loadSpecs(bproDir);

      if (reqs.length === 0) {
        printWarning('No requirements to audit. Run `bpro plan import` or `bpro snapshot` first.');
        return;
      }

      const matrix = loadMatrix(bproDir);
      const entries = matrix?.entries ?? {};
      const root = path.dirname(bproDir);

      const results: Record<string, AuditResult[]> = {
        pass: [],
        warn: [],
        new: [],
        stale: [],
        fail: [],
      };

      for (const req of reqs) {
        if (req.status === 'DEPRECATED') continue;

        const entry = entries[req.id] ?? {};
        const codeRefs = entry.code_refs ?? req.code_refs ?? [];
        const testRefs = entry.test_refs ?? req.test_refs ?? [];

        const hasCode = codeRefs.length > 0;
        let codeExists = false;
        if (hasCode) {
          for (const ref of codeRefs) {
            if (fs.existsSync(path.join(root, ref))) {
              codeExists = true;
              break;
            }
          }
        }

        const hasTest = testRefs.length > 0;

        if (!hasCode) {
          if (['CONFIRMED', 'DRAFT'].includes(req.status)) {
            results.new.push({ id: req.id, title: req.title, detail: 'not implemented' });
          } else {
            results.warn.push({ id: req.id, title: req.title, detail: 'no code mapped' });
          }
        } else if (!codeExists) {
          results.stale.push({ id: req.id, title: req.title, detail: 'code file missing' });
        } else if (!hasTest) {
          results.warn.push({ id: req.id, title: req.title, detail: 'no tests' });
        } else {
          results.pass.push({ id: req.id, title: req.title, detail: 'ok' });
        }
      }

      // If --gate and not --quick, use conductor for deeper analysis
      if (opts.gate && !opts.quick) {
        const config = loadConfig(bproDir);
        const registry = loadModels(bproDir);
        try {
          const adapter = getConductorAdapter(config.conductor, registry);
          const spinner = createSpinner('Running deep audit with conductor...');
          spinner.start();

          const prompt = buildAuditPrompt(reqs);
          const deepResult = await adapter.generateJSON<{
            findings?: Array<{ req_id: string; severity: string; message: string }>;
            gate?: string;
            summary?: string;
          }>(prompt, {
            system: AUDIT_SYSTEM_PROMPT,
            maxTokens: 4096,
          });

          spinner.succeed('Deep audit complete');

          if (deepResult.summary) {
            printInfo(`Conductor: ${deepResult.summary}`);
          }

          appendAgentLog(bproDir, {
            agent: 'auditor',
            action: 'deep-audit',
            model: adapter.name,
            started_at: new Date().toISOString(),
            completed_at: new Date().toISOString(),
            status: 'success',
            output_summary: deepResult.summary ?? 'audit complete',
          });
        } catch {
          printInfo('Deep audit skipped (conductor not available)');
        }
      }

      // Display results
      console.log();

      for (const [, title, detail] of results.pass.map((r) => [r.id, r.title, r.detail])) {
        // skip individual pass display for brevity
      }
      for (const r of results.pass) {
        console.log(`  ${chalk.green('PASS')}  ${r.id}: ${r.title}`);
      }
      for (const r of results.warn) {
        console.log(`  ${chalk.yellow('WARN')}  ${r.id}: ${r.title} — ${r.detail}`);
      }
      for (const r of results.new) {
        console.log(`  ${chalk.blue('TODO')}  ${r.id}: ${r.title} — ${r.detail}`);
      }
      for (const r of results.stale) {
        console.log(`  ${chalk.red('STALE')} ${r.id}: ${r.title} — ${r.detail}`);
      }

      console.log();
      console.log(
        `  PASS ${results.pass.length} | WARN ${results.warn.length} | TODO ${results.new.length} | STALE ${results.stale.length}`,
      );

      // Gate judgment
      if (opts.gate) {
        console.log();
        if (results.stale.length > 0 || results.fail.length > 0) {
          console.log(`  ${chalk.red.bold('Gate: FAIL')} — stale or failing items`);
        } else if (results.warn.length > 0 || results.new.length > 0) {
          console.log(`  ${chalk.yellow.bold('Gate: CONDITIONAL PASS')} — warnings/todos remain`);
        } else {
          console.log(`  ${chalk.green.bold('Gate: PASS')}`);
        }
      }

      // Save audit result
      const ts = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
      const auditResult = {
        timestamp: new Date().toISOString(),
        summary: Object.fromEntries(
          Object.entries(results).map(([k, v]) => [k, v.length]),
        ),
        details: results,
      };
      const reportsDir = path.join(bproDir, 'reports');
      fs.mkdirSync(reportsDir, { recursive: true });
      saveYaml(path.join(reportsDir, `audit-${ts}.yaml`), auditResult);

      console.log();
      console.log(`  ${chalk.dim(`Saved to .bpro/reports/audit-${ts}.yaml`)}`);

      // Emit notification
      const totalIssues = Object.values(auditResult.summary).reduce((a, b) => a + (b as number), 0);
      await emitEvent(bproDir, 'audit.complete', `Audit complete`, {
        Summary: Object.entries(auditResult.summary).map(([k, v]) => `${k}:${v}`).join(' '),
        'Total Items': String(totalIssues),
      });
    } catch (err: unknown) {
      printError(err instanceof Error ? err.message : String(err));
      process.exit(1);
    }
  });
